MapboxMobileEvents/MMEBackgroundLocationServiceTimeoutHandler.m

```diff
@@ -0,0 +1,85 @@
+#import "MMEBackgroundLocationServiceTimeoutHandler.h"
+#import "MMEUIApplicationWrapper.h"
+
+static const NSTimeInterval MMELocationManagerHibernationTimeout = 300.0;
+static const NSTimeInterval MMELocationManagerHibernationPollInterval = 5.0;
+
+@interface MMEBackgroundLocationServiceTimeoutHandler ()
+
+@property (nonatomic) id<MMEUIApplicationWrapper> application;
+@property (nonatomic) NSDate *expiration;
+@property (nonatomic, readwrite) NSTimer *timer;
+@property (nonatomic) UIBackgroundTaskIdentifier backgroundTaskId;
+
+@end
+
+#pragma mark - MMEBackgroundLocationServiceTimeoutHandler
+
+@implementation MMEBackgroundLocationServiceTimeoutHandler
+
+- (instancetype)initWithApplication:(id<MMEUIApplicationWrapper>)application {
+    self = [super init];
+    if (self) {
+        _application = application;
+    }
+    return self;
+}
+
+- (void)timeoutAllowedCheck:(NSTimer *)timer {
+    id<MMEBackgroundLocationServiceTimeoutHandlerDelegate> delegate = self.delegate;
+
+    if (!delegate) {
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [self stopTimer];
+        });
+        return;
+    }
+
+    if (![delegate timeoutHandlerShouldCheckForTimeout:self]) {
```

Comment:
While this is an internal application, would it be good if we check first if the delegate already implements the method?

Reply:
These aren't `optional` protocol methods, so there's no need to check to see if the delegate already implements it. The compiler will warn if the client hasn't implemented them.

Since these are really internal methods, I didn't see a need to make them optional.